jobwatcher/plugins/sge.py

```diff
@@ -0,0 +1,52 @@
+import subprocess as sub
+import os
+import shlex
+import logging
+import json
+
+log = logging.getLogger(__name__)
+
+# get nodes requested from pending jobs
+def get_required_nodes(instance_properties):
+    command = "qstat -g d -s p -u '*'"
+    _output = run_command(command)
+    slots = 0
+    output = _output.split("\n")[2:]
+    for line in output:
+        line_arr = line.split()
+        if len(line_arr) >= 8:
+            slots += int(line_arr[7])
+    vcpus = instance_properties.get('slots')
+    return -(-slots // vcpus)
+
+# get nodes reserved by running jobs
+# if a host has 1 or more job running on it, it'll be marked busy
+def get_busy_nodes(instance_properties):
+    command = "qstat -f"
+    _output = run_command(command)
+    nodes = 0
+    output = _output.split("\n")[2:]
+    for line in output:
+        line_arr = line.split()
+        if len(line_arr) == 5:
+            # resv/used/tot.
+            (resv, used, total) = line_arr[2].split('/')
+            if int(used) > 0 or int(resv) > 0:
+                nodes += 1
+    return nodes
+
+def run_command(command):
+    _command = shlex.split(command)
+    try:
+        DEV_NULL = open(os.devnull, "rb")
+        process = sub.Popen(_command, env=dict(os.environ), stdout=sub.PIPE, stderr=sub.STDOUT, stdin=DEV_NULL)
+        _output = process.communicate()[0]
+        exitcode = process.poll()
+        if exitcode != 0:
```

Comment:
This could be an issue if there is nothing written to stdout. I think in such cases exitcode is almost always 1. Could you please verify that?

Reply:
that's just for grep. For qstat, it returns a 0 exit code even with no output.